libcontainer/init_linux.go

```diff
@@ -131,6 +131,25 @@ func finalizeNamespace(config *initConfig) error {
 		return errors.Wrap(err, "close exec fds")
 	}
 
+	// we only do chdir if it's specified
+	doChdir := config.Cwd != ""
+	if doChdir {
+		doChdir = false
```

Comment:
I'd rather set doChdir to false after `unix.Chdir` has succeeded (and modify the comment below saying we ignore EPERM for now because we'll retry below).

Not too important though, it is fine as it is.

Reply:
I like the code flow here better than the alternative of :
```
	if doChdir {
		// First, attempt the chdir before setting up the user.
		// This could allow us to access a directory that the user running runc can access
		// but the container user cannot.
		if err := unix.Chdir(config.Cwd); err != nil {
			if err != unix.EPERM {
				return fmt.Errorf("chdir to cwd (%q) set in config.json failed: %v", config.Cwd, err)
			}
			// If we hit an EPERM, we should attempt again after setting up user.
			// This will allow us to successfully chdir if the container user has access
			// to the directory, but the user running runc does not.
			// This is useful in cases where the cwd is also a volume that's been chowned to the container user.
		} else {
			doChdir = false
		}
	}
```
(elses are kinda ugly imo)

Reply:
You can do switch/case, looks decent to me:

```go
switch err := unix.Chdir(config.Cwd); err {
case nil: 
    doChdir = false
case EPERM:
    // will retry below
default:
    return fmt.Errorf("chdir to cwd (%q) set in config.json failed: %v", config.Cwd, err)
}
```

Reply:
agreed, that's cleaner. I've adopted it (but can't switch on err because of the failures caused by not using IsPermission)